File: tree-walk/cs/Lox/Interpreter/LoxInstance.cs
﻿using System;
using Lox.Scanner;
using Lox.Interpreter.Callables;

namespace Lox.Interpreter;

class LoxInstance
{
	private readonly LoxClass klass;
	private readonly Dictionary<string, object?> fields = new Dictionary<string, object?>();

	public LoxInstance(LoxClass klass)
	{
		this.klass = klass;
	}

	public object? Get(Token name)
	{
		if (fields.ContainsKey(name.Lexeme))
		{
			return fields[name.Lexeme];
		}

		var method = klass.FindMethod(name.Lexeme);
		if (method != null)
		{
			return method.Bind(this);
		}

		throw new RuntimeError(name, $"Undefined property '{name.Lexeme}'.");
	}

	public void Set(Token name, object? value)
	{
		fields[name.Lexeme] = value;
	}

    public override string ToString()
    {
		return $"{klass.Name} instance";
    }
}

File: tree-walk/go/internal/lox/lox_instance.go
package lox

import (
	"fmt"
)

type loxInstance struct {
	class  *loxClass
	fields map[string]any
}

func newLoxInstance(class *loxClass) *loxInstance {
	return &loxInstance{
		class:  class,
		fields: make(map[string]any),
	}
}

func (i *loxInstance) get(name *Token) any {
	if value, ok := i.fields[name.Lexeme]; ok {
		return value
	}

	method := i.class.findMethod(name.Lexeme)
	if method != nil {
		return method.bind(i)
	}

	panic(newRuntimeError(name, fmt.Sprintf("Undefined property '%s'.", name.Lexeme)))
}

func (i *loxInstance) set(name *Token, value any) {
	i.fields[name.Lexeme] = value
}

func (i *loxInstance) String() string {
	return i.class.name + " instance"
}
